## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [85]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv',dtype=None)

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [86]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [87]:
# Use the describe function to gather some basic statistics

df = df.astype({"order_id": str}, errors='raise')
df = df.astype({"qty": int}, errors='raise')

df.describe()

client_id    order_week    order_year    unit_price     unit_cost  \
count  54639.000000  54639.000000  54639.000000  54639.000000  54639.000000   
mean   54837.869416     11.359139   2022.993064    136.267207     99.446073   
std    25487.438231      7.023499      0.082997    183.873135    133.164267   
min    10033.000000      1.000000   2022.000000      0.010000      0.010000   
25%    33593.000000      6.000000   2023.000000     20.800000     14.840000   
50%    53305.000000     11.000000   2023.000000     68.310000     49.890000   
75%    78498.000000     17.000000   2023.000000    173.160000    125.570000   
max    99984.000000     52.000000   2023.000000   1396.230000    846.270000   

        unit_weight           qty   line_number  
count  54639.000000  5.463900e+04  54639.000000  
mean       5.004116  5.702646e+02      2.979667  
std        5.326599  1.879552e+04      2.436320  
min        0.000000  0.000000e+00      0.000000  
25%        1.450000  3.200000e+01      1.000000  
50%        3.240000  6.800000e+01      3.000000  
75%        6.890000  1.700000e+02      5.000000  
max       46.430000  3.958244e+06      9.000000

In [88]:
# Use this space to do any additional research
# and familiarize yourself with the data.

# For the category with the most entries,
# which subcategory had the most entries?

# sum(df['qty'])

df.shape

(54639, 18)

In [89]:
# What three item categories had the most entries?
# consumables, furniture, software
df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [90]:
# For the category with the most entries,
# which subcategory had the most entries?
is_consumable = df['category'] == 'consumables'
df.loc[is_consumable]['subcategory'].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [91]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head(5)



client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [92]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = list(df['client_id'].value_counts().head(5).keys())

top_5_clients

[33615, 66037, 46820, 38378, 24741]

In [93]:
# How many total units (the qty column) did the
# client with the most entries order order?
# 33615
most_entries = df[['client_id']] == top_5_clients[0]
top_client = df.loc[most_entries["client_id"],:]
sum(top_client['qty'])


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [94]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

#top_client.loc[:,'line_subtotal'] = top_client.apply(lambda x: x.qty * x.unit_price, axis=1)
#top_client
df.loc[:,'line_subtotal'] = df.apply(lambda x: x.qty * x.unit_price, axis=1)
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515  8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609  8069089  2023-05-19   
2                    swatson8146@payne.net      57113  1902144  2023-01-29   
3                   bakernathan@benson.com      46554  9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089  1322274  2023-05-28   
...                                    ...        ...      ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043  9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908  6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176  8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485  7592730  2023-01-18   
54638                     belld@galvan.com      24007  7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  line_subtotal  
0                1      115164.00  
1                0         523.95  
2                6         527.28  
3                3        1056.18  
4                1        3902.00  
...            ...            ...  
54634            8        2743.29  
54635            0        9709.73  
54636            6       31188.50  
54637            0         165.76  
54638            5        3103.31  

[54639 rows x 19 columns]

In [95]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

def calc_shipping(row):
    pplb = 7
    if row.unit_weight <= 50:
        pplb = 10
    
    return pplb * row.unit_weight


# top_client.loc[:,'shipping_price'] = top_client.apply(calc_shipping, axis=1)
# top_client

df.loc[:,'shipping_price'] = df.apply(calc_shipping, axis=1)
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515  8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609  8069089  2023-05-19   
2                    swatson8146@payne.net      57113  1902144  2023-01-29   
3                   bakernathan@benson.com      46554  9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089  1322274  2023-05-28   
...                                    ...        ...      ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043  9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908  6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176  8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485  7592730  2023-01-18   
54638                     belld@galvan.com      24007  7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  line_subtotal  shipping_price  
0                1      115164.00            75.0  
1                0         523.95            14.9  
2                6         527.28            16.8  
3                3        1056.18            12.3  
4                1        3902.00           464.3  
...            ...            ...             ...  
54634            8        2743.29            22.5  
54635            0        9709.73           117.0  
54636            6       31188.50            41.6  
54637            0         165.76          

In [96]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

# top_client.loc[:,'line_total_price'] = top_client.apply(lambda x: x.line_subtotal + x.shipping_price + (x.line_subtotal*.0925), axis=1)
# top_client

df.loc[:,'line_total_price'] = df.apply(lambda x: ( (x.line_subtotal + x.shipping_price) *1.0925), axis=1)
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515  8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609  8069089  2023-05-19   
2                    swatson8146@payne.net      57113  1902144  2023-01-29   
3                   bakernathan@benson.com      46554  9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089  1322274  2023-05-28   
...                                    ...        ...      ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043  9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908  6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176  8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485  7592730  2023-01-18   
54638                     belld@galvan.com      24007  7489403  2023-01-04   

       order_week  order_year  ...     category         subcategory  \
0              17        2023  ...        decor            wall art   
1              20        2023  ...  consumables                pens   
2               4        2023  ...     software  project management   
3              17        2023  ...  consumables                pens   
4              21        2023  ...  consumables                misc   
...           ...         ...  ...          ...                 ...   
54634          17        2023  ...     software           logistics   
54635           6        2023  ...  consumables    kitchen supplies   
54636           9        2023  ...  consumables   bathroom supplies   
54637           3        2023  ...    furniture              stands   
54638           1        2023  ...     software  project management   

      unit_price  unit_cost  unit_weight   qty  line_number  line_subtotal  \
0        1096.80     762.71         7.50   105            1      115164.00   
1          24.95      15.09         1.49    21            0         523.95   
2          13.52       7.86         1.68    39            6         527.28   
3          36.42      24.85         1.23    29            3        1056.18   
4         195.10     108.17        46.43    20            1        3902.00   
...          ...        ...          ...   ...          ...            ...   
54634      83.13      51.60         2.25    33            8        2743.29   
54635     206.59     175.46        11.70    47            0        9709.73   
54636      65.66      57.31         4.16   475            6       31188.50   
54637       1.48       1.22        18.04   112            0         165.76   
54638       3.01       1.61         2.07  1031            5        3103.31   

       shipping_price  line_total_price  
0                75.0     125898.607500  
1                14.9        588.693625  
2                16.8        594.407400  
3                12.3       1167.314400  
4               464.3       4770.182750  
...               ...               ...  
54634            22.5       3021.625575  
54635           117.0      10735.70

In [97]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

# top_client.loc[:,'line_cost'] = top_client.apply(lambda x: (x.qty*x.unit_cost) + x.shipping_price, axis=1)

# top_client

df.loc[:,'line_cost'] = df.apply(lambda x: (x.qty*x.unit_cost) + x.shipping_price, axis=1)
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515  8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609  8069089  2023-05-19   
2                    swatson8146@payne.net      57113  1902144  2023-01-29   
3                   bakernathan@benson.com      46554  9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089  1322274  2023-05-28   
...                                    ...        ...      ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043  9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908  6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176  8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485  7592730  2023-01-18   
54638                     belld@galvan.com      24007  7489403  2023-01-04   

       order_week  order_year  ...         subcategory unit_price unit_cost  \
0              17        2023  ...            wall art    1096.80    762.71   
1              20        2023  ...                pens      24.95     15.09   
2               4        2023  ...  project management      13.52      7.86   
3              17        2023  ...                pens      36.42     24.85   
4              21        2023  ...                misc     195.10    108.17   
...           ...         ...  ...                 ...        ...       ...   
54634          17        2023  ...           logistics      83.13     51.60   
54635           6        2023  ...    kitchen supplies     206.59    175.46   
54636           9        2023  ...   bathroom supplies      65.66     57.31   
54637           3        2023  ...              stands       1.48      1.22   
54638           1        2023  ...  project management       3.01      1.61   

       unit_weight   qty  line_number  line_subtotal  shipping_price  \
0             7.50   105            1      115164.00            75.0   
1             1.49    21            0         523.95            14.9   
2             1.68    39            6         527.28            16.8   
3             1.23    29            3        1056.18            12.3   
4            46.43    20            1        3902.00           464.3   
...            ...   ...          ...            ...             ...   
54634         2.25    33            8        2743.29            22.5   
54635        11.70    47            0        9709.73           117.0   
54636         4.16   475            6       31188.50            41.6   
54637        18.04   112            0         165.76           180.4   
54638         2.07  1031            5        3103.31            20.7   

       line_total_price  line_cost  
0         125898.607500   80159.55  
1            588.693625     331.79  
2            594.407400     323.34  
3           1167.314400     732.95  
4           4770.182750    2627.70  
...                 ...        ...  
54634       3021.625575    1725.30  
54635      10735.702525    8363.62  
54636      341

In [98]:
# Create a column for the profit of
# each line using line cost and line price

# top_client.loc[:,'line_profit'] = top_client.apply(lambda x: x.line_total_price - x.line_cost, axis=1)
# top_client

df.loc[:,'line_profit'] = df.apply(lambda x: x.line_total_price - x.line_cost, axis=1)
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515  8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609  8069089  2023-05-19   
2                    swatson8146@payne.net      57113  1902144  2023-01-29   
3                   bakernathan@benson.com      46554  9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089  1322274  2023-05-28   
...                                    ...        ...      ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043  9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908  6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176  8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485  7592730  2023-01-18   
54638                     belld@galvan.com      24007  7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number  line_subtotal  shipping_price  line_total_price  \
0                1      115164.00            75.0     125898.607500   
1                0         523.95            14.9        588.693625   
2                6         527.28            16.8        594.407400   
3                3        1056.18            12.3       1167.314400   
4                1        3902.00           464.3       4770.182750   
...            ...            ...             ...               ...   
54634            8        2743.29            22.5       3021.625575   
54635            0        9709.73           117.0      10735.702525   
54636            6       31188.50            41.6      34118.884250   
54637            0         165.76           180.4        378.179800   
54638            5        3103.31            20.7       3412.980925   

       line_cost   line_profit  
0       80159.55  45739.057500  
1         331.79    256.903625  
2         323.34    271.067400  
3         732.95    434.364400  
4        2627.70   2142.482750  
...          ...           ...  
54634    1725.30   1296.325575  
54635    8363.62   2372.082525  
54636   27263.85   6855.034250  
54637     317.04     61.139800  
54638    1680.61   1

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [101]:
# Check your work using the totals above
grouped_df = df.set_index('order_id')
grouped_df = df.set_index('order_id').loc[['2742071','2173913','6128929'],['line_total_price','line_profit']].groupby(by='order_id').sum()

#grouped_df.loc['2742071',['unit_price', 'unit_cost', 'unit_weight', 'qty', 'line_number', 'line_subtotal', 'shipping_price', 'line_total_price', 'line_cost', 'line_profit']]

#grouped_df.loc[['2742071','2173913','6128929'],'line_total_price'].sum()
grouped_df

#answer
# Order 2742071 total: $152811.89
# Order 2173913 total: $162388.71
# Order 6128929 total: $923441.25

line_total_price    line_profit
order_id                                 
2173913      119838.357050   28285.887050
2742071      139844.512025   27655.962025
6128929      752748.745475  283248.305475

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.



33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [17]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.



client_id     qty  shipping_price   line_price    line_cost  line_profit
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61

In [18]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.



In [19]:
# Sort the updated data by "Total Profit" form highest to lowest


Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M

In [ ]:
def format_currency(amount):
    return '${:,.2f}'.format(amount)
#then [some number].map(format_currency)